patches/chromium/feat_ensure_mas_builds_of_the_same_application_can_use_safestorage.patch

```diff
@@ -0,0 +1,105 @@
+ using KeychainNameContainerType = const base::NoDestructor<std::string>;
+ #endif
+ 
++#if defined(MAS_BUILD)
++const char kAccountNameSuffix[] = " App Store Key";
++#else
++const char kAccountNameSuffix[] = " Key";
++#endif
++
+ namespace {
+ 
+ // These two strings ARE indeed user facing.  But they are used to access
+@@ -82,11 +88,18 @@
+ std::string KeychainPassword::GetPassword() const {
+   UInt32 password_length = 0;
+   void* password_data = nullptr;
++  KeychainPassword::KeychainNameType service_name = GetServiceName();
++  KeychainPassword::KeychainNameType account_name = GetAccountName();
++  const std::string account_name_suffix = kAccountNameSuffix;
++  const std::string suffixed_account_name = account_name + account_name_suffix;
++
++  // We should check if the suffixed account exists first
+   OSStatus error = keychain_.FindGenericPassword(
+-      GetServiceName().size(), GetServiceName().c_str(),
+-      GetAccountName().size(), GetAccountName().c_str(), &password_length,
++      service_name.size(), service_name.c_str(),
++      suffixed_account_name.size(), suffixed_account_name.c_str(), &password_length,
+       &password_data, nullptr);
+ 
++  // If it exists we can return it immediately
+   if (error == noErr) {
+     std::string password =
+         std::string(static_cast<char*>(password_data), password_length);
+@@ -94,9 +107,52 @@
+     return password;
+   }
+ 
++  // If the error was anything other than "it does not exist" we should error out here
++  // This normally means the account exists but we were deniged access to it
++  if (error != errSecItemNotFound) {
++    OSSTATUS_LOG(ERROR, error) << "Keychain lookup for suffixed key failed";
++    return std::string();
++  }
++
++  // If the suffixed account didn't exist, we should check if the legacy non-suffixed account
++  // exists. If it does we can use that key and migrate it to the new account
++  base::ScopedCFTypeRef<SecKeychainItemRef> item_ref;
++  error = keychain_.FindGenericPassword(
++      service_name.size(), service_name.c_str(),
++      account_name.size(), account_name.c_str(), &password_length,
++      &password_data, item_ref.InitializeInto());
++
++  if (error == noErr) {
++    std::string password =
++        std::string(static_cast<char*>(password_data), password_length);
++
++    // If we found the legacy account name we should copy it over to
++    // the new suffixed account
++    error = keychain_.AddGenericPassword(
++        service_name.size(), service_name.data(), suffixed_account_name.size(),
++        suffixed_account_name.data(), password.size(), password_data, NULL);
++
++    if (error == noErr) {
++      // If we successfully made the suffixed account we can delete the old
++      // account to ensure new apps don't try to use it and run into IAM
++      // issues
++      error = keychain_.ItemDelete(item_ref.get());
++      if (error != noErr) {
++        OSSTATUS_DLOG(ERROR, error) << "Keychain delete for legacy key failed";
```

Comment:
Not sure when this case happens, but if we fail to delete the old account and successfully copy over the new account, seems like successive access would no longer hit this path to delete the old account. Do we leave the old account around in that case ?

Reply:
Yeah, I think that's ok.  The alternative is we try to delete it on successful reads of the suffixed key.  But deleting it requires a lookup and that could throw up a unnecessary permission prompt.  I'd rather just leave the old key there.  I don't think a delete failure is likely based on my understanding of the keychain